actors/builtin/miner/miner_test.go

```diff
@@ -2280,6 +2327,105 @@ func TestWindowPost(t *testing.T) {
+		dlIdx, pIdx, err := st.FindSector(store, sector.SectorNumber)
+		require.NoError(t, err)
+
+		// Skip over deadlines until the beginning of the one with the new sector
+		dlinfo := actor.deadline(rt)
+		for dlinfo.Index != dlIdx {
+			dlinfo = advanceDeadline(rt, actor, &cronConfig{})
+		}
+
+		// Submit PoSt
+		partitions := []miner.PoStPartition{
+			{Index: pIdx, Skipped: bitfield.New()},
+		}
+		actor.submitWindowPoStRaw(rt, dlinfo, partitions, []*miner.SectorOnChainInfo{sector}, proofs, &poStConfig{
+			expectedPowerDelta: pwr,
+		})
+
+		// Dispute it.
+		params := miner.DisputeWindowedPoStParams{
+			Deadline:  dlinfo.Index,
+			PoStIndex: 0,
+		}
+
+		rt.SetCaller(actor.worker, builtin.AccountActorCodeID)
+		rt.ExpectValidateCallerType(builtin.CallerTypesSignable...)
+
+		expectQueryNetworkInfo(rt, actor)
+
+		rt.ExpectAbortContainsMessage(exitcode.ErrForbidden, "can only dispute window posts during the dispute window", func() {
+			rt.Call(actor.a.DisputeWindowedPoSt, &params)
+		})
+		rt.Verify()
+	})
+	t.Run("can dispute up till window end, but not after", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+		store := rt.AdtStore()
+		sector := actor.commitAndProveSectors(rt, 1, defaultSectorExpiration, nil)[0]
+
+		st := getState(rt)
+		dlIdx, _, err := st.FindSector(store, sector.SectorNumber)
+		require.NoError(t, err)
+
+		nextDl := miner.NewDeadlineInfo(st.ProvingPeriodStart, dlIdx, rt.Epoch()).
+			NextNotElapsed()
+
+		advanceAndSubmitPoSts(rt, actor, sector)
+
+		windowEnd := nextDl.Close + miner.WPoStDisputeWindow
+
+		// first, try to dispute right before the window end.
+		// We expect this to fail "normally" (fail to disprove).
+		rt.SetEpoch(windowEnd - 1)
+		actor.disputeWindowPoSt(rt, nextDl, 0, []*miner.SectorOnChainInfo{sector}, nil)
+
+		// Now set the epoch at the window end. We expect a different error.
+		rt.SetEpoch(windowEnd)
+
+		// Now try to dispute.
+		params := miner.DisputeWindowedPoStParams{
+			Deadline:  dlIdx,
+			PoStIndex: 0,
+		}
+
+		rt.SetCaller(actor.worker, builtin.AccountActorCodeID)
+		rt.ExpectValidateCallerType(builtin.CallerTypesSignable...)
+
+		currentReward := reward.ThisEpochRewardReturn{
+			ThisEpochBaselinePower:  actor.baselinePower,
+			ThisEpochRewardSmoothed: actor.epochRewardSmooth,
+		}
+		rt.ExpectSend(builtin.RewardActorAddr, builtin.MethodsReward.ThisEpochReward, nil, big.Zero(), &currentReward, exitcode.Ok)
+
+		networkPower := big.NewIntUnsigned(1 << 50)
+		rt.ExpectSend(builtin.StoragePowerActorAddr, builtin.MethodsPower.CurrentTotalPower, nil, big.Zero(),
+			&power.CurrentTotalPowerReturn{
+				RawBytePower:            networkPower,
+				QualityAdjPower:         networkPower,
+				PledgeCollateral:        actor.networkPledge,
+				QualityAdjPowerSmoothed: actor.epochQAPowerSmooth,
+			},
+			exitcode.Ok)
+
+		rt.ExpectAbortContainsMessage(exitcode.ErrForbidden, "can only dispute window posts during the dispute window", func() {
```

Comment:
Nit: you should be able to call `actor.disputeWindowPoSt` inside `rt.ExpectAbortContainsMessage`, and I would expect that to encapsulate all the expectation set-up above that you had to replicate. `Reset()` the mock afterwards some some expectations won't have been reached.

Reply:
Actually, this is a bit of a hole in our tests that we should probably fix. I'm trying to test that the expectations are met, even on the failure case. Otherwise, we can get gas mismatches on failure cases.

Reply:
Fair. I hope that some kind of golden file approach in the future will abstract this away. Your call in the meantime.

Reply:
Yeah, I agree that golden files will make this kind of thing unnecessary.